File: bits.go
// Functions for doing various bit conversions/manipulations.

package main

func subBits(bits [8]int, start int, stop int) [8]int {
    var output [8]int
    var pos = 0

    for i := start; i < stop; i++ {
        output[pos] = bits[i]
    }

    return output
}

func numberToBits(number int) [8]int {
    var pos int = 7
    var output [8]int

    for divider := 128; divider >= 1; divider=divider/2 {
        testNumber := number/divider

        if testNumber >= 1 {
            number = number - divider
            output[pos] = 1
        } else {
            output[pos] = 0
        }

        pos--
    }

    return output
}

func bitsToNumber(bits [8]int) int {
    var output int = 0
    var pos int = 0

    for multiplier := 1; multiplier < 256; multiplier *= 2 {
        if bits[pos] == 1 {
            output += multiplier
        }

        pos++
    }

    return output
}

func charToBits(char string) [8]int {
    return numberToBits(int(char[0]))
}

func shiftBits(bits [8]int, offset int) [8]int {
    var output [8]int

    for x := 0; x < 8; x++ {
        destination := x + offset
        if destination > 7 {
            destination -= 8
        }
        if destination < 0 {
            destination += 8
        }

        output[destination] = bits[x]
    }

    return output
}

File: prototype/python/bits.py
import numpy

class Bits():
    def __init__(self):
        pass

    def charToBits(self, char):
        return self.numberToBits(ord(char))

    def numberToBits(self, number):
        pos = 7
        output = numpy.array([0, 0, 0, 0, 0, 0, 0, 0])
        dividor = 128

        while dividor > 0.5:
            testNumber = number / dividor

            if testNumber >= 1:
                number -= dividor
                output[pos] = 1

            dividor /= 2
            pos -= 1

        return output

    def bitsToNumber(self, bits):
        pos = 0
        output = 0
        multiplier = 1
        inputLength = len(bits)

        while multiplier < 256:
            if not pos < inputLength:
                break

            if bits[pos] == 1:
                output += multiplier

            pos += 1
            multiplier *= 2


        return output

    def bitsToChar(self, bits):
        return chr(self.bitsToNumber(bits))

    def shiftBits(self, bits, shiftBy):
        output = numpy.array([0, 0, 0, 0, 0, 0, 0, 0])

        for pos in range(0, 8):
            outPos = pos + shiftBy

            if outPos > 7:
                outPos -= 8

            if outPos < 0:
                outPos += 8

            output[outPos] = bits[pos]

        return output

    def subBits(self, bits, start, stop):
        stopBoundary = stop + 1
        output = []
        numberOfBits = len(bits)


        for pos in range(start, stopBoundary):
            if pos >= numberOfBits:
                break

            output.append(bits[pos])

        return output
